docs/config_file.md
## Config file (ini)

```ini
work_dir_path = ./ # work dir path usually is /var/sniffdogsniff
log_to_file = no
log_file_name = sds.log
search_database_max_ram_cache_size = 512M # max cache in ram before flush to disk
allow_results_invalidation = no 
min_search_results_threshold = 50                 # minimum results threshold for sds to search on other engines (if there is no 50 results found in the database then sds search on centralized engines)
```

### Initial Peers
```ini
[peer]
address = 127.0.0.1:4222
id = 199ee773164f7a8bedcbef3afa1bf398cf5febeb

[peer]
address = somepeer.onion
id = 1ec6c7a7ec625077c958e9a321ec7f20c7725e70
```
The initial peers to let you jump into the p2p network. Sniffdogsniff uses kademlia
so the id of a peer is required. Usually is the sha1 of the address.
Can be added as much initial peers as wanted by adding peer section multiple times.

#### Proxy Settings
```ini
[proxy_settings]
i2p_socks5_proxy = addr:port
tor_socks5_proxy = 127.0.0.1:9050
```

* *i2p_socks5_proxy* is the address of the i2p or i2pd daemon socks proxy on your machine
* *tor_socks5_proxy* is the address of the tor daemon socks proxy on your machine

### Node RPC p2p server
```ini
[node_service]
enabled = yes # node is visible to other nodes
bind_address = 0.0.0.0:4222
```
If enabled allows other nodes to sync with your node. Bind address specify the p2p
bind address. If using tor or i2p service auto creation bind_address key is not 
needed.

* Example of configuration for tor onionservice listen address auto configuration
```ini
[node_service]
enabled = yes
hidden_service = tor
tor_control_port = 9051
tor_control_auth_password = password
```

* Example of configuration for i2p hiddenservice listen address auto configuration
```ini
[node_service]
enabled = yes # node is visible to other nodes
hidden_service = i2p
i2p_sam_port = 7656
i2p_sam_user = user
i2p_sam_password = password
```

### Web Ui
```ini
[web_ui]
bind_address = 0.0.0.0:8081 # bind address for web server
```



core/config.go
package core

import (
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"

	"github.com/sniffdogsniff/hiddenservice"
	"github.com/sniffdogsniff/kademlia"
	"github.com/sniffdogsniff/logging"
	"github.com/sniffdogsniff/proxies"
	"gopkg.in/ini.v1"
)

const MAX_RAM_DB_SIZE = 268435456 // 256 MB

const (
	DEFAULT_LOG_FILE_NAME = "sds.log"
)

const (
	SERVICE_RPC_PORT                   = "service_rpc_port"
	WORK_DIR_PATH                      = "work_dir_path"
	LOG_FILE_NAME                      = "log_file_name"
	LOG_TO_FILE                        = "log_to_file"
	ALLOW_RESULTS_INVALIDATION         = "allow_results_invalidation"
	SEARCH_DATABASE_MAX_RAM_CACHE_SIZE = "search_database_max_ram_cache_size"
	PEER                               = "peer"
	ID                                 = "id"
	ADDRESS                            = "address"
	EXTERNAL_SEARCH_ENGINE             = "external_search_engine"
	NAME                               = "name"
	USER_AGENT                         = "user_agent"
	SEARCH_QUERY_URL                   = "search_query_url"
	RESULTS_CONTAINER_ELEMENT          = "results_container_element"
	RESULT_CONTAINER_ELEMENT           = "result_container_element"
	RESULT_URL_ELEMENT                 = "result_url_element"
	RESULT_URL_PROPERTY                = "result_url_property"
	RESULT_URL_IS_JSON                 = "result_url_is_json"
	RESULT_URL_JSON_PROPERTY           = "result_url_json_property"
	RESULT_TITLE_ELEMENT               = "result_title_element"
	RESULT_TITLE_PROPERTY              = "result_title_property"
	PROVIDED_DATA_TYPE                 = "provided_data_type"
	WEB_UI                             = "web_ui"
	BIND_ADDRESS                       = "bind_address"
	PROXY_SETTINGS                     = "proxy_settings"
	FORCE_TOR_PROXY                    = "force_tor_proxy"
	I2P_SOCKS5_PROXY                   = "i2p_socks5_proxy"
	TOR_SOCKS5_PROXY                   = "tor_socks5_proxy"
	NODE_SERVICE                       = "node_service"
	ENABLED                            = "enabled"
	HIDDEN_SERVICE                     = "hidden_service"
	TOR_CONTROL_PORT                   = "tor_control_port"
	TOR_CONTROL_AUTH_PASSWORD          = "tor_control_auth_password"
	I2P_SAM_PORT                       = "i2p_sam_port"
	I2P_SAM_USER                       = "i2p_sam_user"
	I2P_SAM_PASSWORD                   = "i2p_sam_password"
)

func panicNoKey(key string) {
	panic(fmt.Sprint("Config file parse error: required key: ", key))
}

func panicNoSection(key string) {
	panic(fmt.Sprint("Config file parse error: required section: ", key))
}

func stringToByteSize(text string) int {
	cleanStr := strings.Trim(text, " ")
	size, err := strconv.Atoi(cleanStr[0 : len(cleanStr)-2])
	if err != nil {
		logging.LogWarn("Cannot parse db cache size")
		return MAX_RAM_DB_SIZE
	}
	if strings.HasSuffix(cleanStr, "G") { // Gigs
		return size * 1024 * 1024 * 1024
	} else if strings.HasSuffix(cleanStr, "M") { // Megs
		return size * 1024 * 1024
	} else if strings.HasSuffix(cleanStr, "K") { // Kilos
		return size * 1024
	} else { // Bytes
		return size
	}
}

func StrToDataType(token string) ResultDataType {
	switch token {
	case "images":
		return IMAGE_DATA_TYPE
	case "videos":
		return VIDEO_DATA_TYPE
	case "links":
		return LINK_DATA_TYPE
	default:
		return LINK_DATA_TYPE
	}
}

type SdsConfig struct {
	WorkDirPath              string
	LogFileName              string
	LogToFile                bool
	searchDBMaxCacheSize     int
	AllowResultsInvalidation bool
	WebServiceBindAddress    string
	KnownPeers               map[kademlia.KadId]string
	ProxySettings            proxies.ProxySettings
	P2PServerEnabled         bool
	P2PServerProto           hiddenservice.NetProtocol
	searchEngines            map[string]SearchEngine
}

func NewSdsConfig(path string) SdsConfig {
	cfg := SdsConfig{}
	iniData, err := ini.LoadSources(ini.LoadOptions{
		AllowNonUniqueSections: true,
	}, path)

	if err != nil {
		panic(err.Error())
	}

	defaultSection := iniData.Section(ini.DefaultSection)
	if defaultSection.HasKey(WORK_DIR_PATH) {
		cfg.WorkDirPath = defaultSection.Key(WORK_DIR_PATH).String()
	} else {
		panicNoKey(WORK_DIR_PATH)
	}
	if defaultSection.HasKey(LOG_TO_FILE) {
		cfg.LogToFile = defaultSection.Key(LOG_TO_FILE).MustBool(false)
	} else {
		cfg.LogToFile = false
	}
	if defaultSection.HasKey(LOG_FILE_NAME) {
		cfg.LogFileName = defaultSection.Key(LOG_TO_FILE).String()
	} else {
		cfg.LogFileName = DEFAULT_LOG_FILE_NAME
	}
	if defaultSection.HasKey(SEARCH_DATABASE_MAX_RAM_CACHE_SIZE) {
		cfg.searchDBMaxCacheSize = stringToByteSize(
			defaultSection.Key(SEARCH_DATABASE_MAX_RAM_CACHE_SIZE).String())
	} else {
		cfg.searchDBMaxCacheSize = MAX_RAM_DB_SIZE
	}
	if defaultSection.HasKey(ALLOW_RESULTS_INVALIDATION) {
		allowInvalidation, err := defaultSection.Key(ALLOW_RESULTS_INVALIDATION).Bool()
		if err != nil {
			cfg.AllowResultsInvalidation = allowInvalidation
		} else {
			cfg.AllowResultsInvalidation = false
		}
	} else {
		cfg.AllowResultsInvalidation = false
	}

	cfg.KnownPeers = make(map[kademlia.KadId]string)
	peersSections, err := iniData.SectionsByName(PEER)
	if err != nil {
		panicNoSection(PEER)
	}
	for _, pSec := range peersSections {
		logging.LogTrace("read.config.ini")
		if pSec.HasKey(ADDRESS) && pSec.HasKey(ID) {
			idBytez, err := hex.DecodeString(pSec.Key(ID).String())
			if err != nil {
				continue
			}
			kadId := kademlia.KadIdFromBytes(idBytez)
			cfg.KnownPeers[kadId] = pSec.Key(ADDRESS).String()
		}
	}

	cfg.searchEngines = make(map[string]SearchEngine)
	enginesSections, err := iniData.SectionsByName(EXTERNAL_SEARCH_ENGINE)
	if err != nil {
		logging.LogWarn("Failed to parse config file: external search engines")
	}
	for _, sec := range enginesSections {
		name := sec.Key(NAME).String()
		cfg.searchEngines[name] = SearchEngine{
			name:                    name,
			userAgent:               sec.Key(USER_AGENT).String(),
			searchQueryUrl:          sec.Key(SEARCH_QUERY_URL).String(),
			resultsContainerElement: sec.Key(RESULTS_CONTAINER_ELEMENT).String(),
			resultContainerElement:  sec.Key(RESULT_CONTAINER_ELEMENT).String(),
			resultUrlElement:        sec.Key(RESULT_URL_ELEMENT).String(),
			resultUrlProperty:       sec.Key(RESULT_URL_PROPERTY).String(),
			resultUrlIsJson:         sec.Key(RESULT_URL_IS_JSON).MustBool(false),
			resultUrlJsonProperty:   sec.Key(RESULT_URL_JSON_PROPERTY).String(),
			resultTitleElement:      sec.Key(RESULT_TITLE_ELEMENT).String(),
			resultTitleProperty:     sec.Key(RESULT_TITLE_PROPERTY).String(),
			providedDataType:        StrToDataType(sec.Key(PROVIDED_DATA_TYPE).String()),
		}
	}

	if iniData.HasSection(WEB_UI) {
		webUiSection := iniData.Section(WEB_UI)
		if webUiSection.HasKey(BIND_ADDRESS) {
			cfg.WebServiceBindAddress = webUiSection.Key(BIND_ADDRESS).String()
		} else {
			cfg.WebServiceBindAddress = "127.0.0.1:8081"
		}
	} else {
		panicNoSection(WEB_UI)
	}

	if iniData.HasSection(PROXY_SETTINGS) {
		proxySettingsSection := iniData.Section(PROXY_SETTINGS)
		if proxySettingsSection.HasKey(FORCE_TOR_PROXY) {
			cfg.ProxySettings.ForceTor, err = proxySettingsSection.Key(FORCE_TOR_PROXY).Bool()
			if err != nil {
				cfg.ProxySettings.ForceTor = false
			}
		} else {
			cfg.ProxySettings.ForceTor = false
		}
		if proxySettingsSection.HasKey(TOR_SOCKS5_PROXY) {
			cfg.ProxySettings.TorSocks5Addr = proxySettingsSection.Key(TOR_SOCKS5_PROXY).String()
		} else {
			cfg.ProxySettings.TorSocks5Addr = "127.0.0.1:9050"
		}
		if proxySettingsSection.HasKey(I2P_SOCKS5_PROXY) {
			cfg.ProxySettings.I2pSocks5Addr = proxySettingsSection.Key(I2P_SOCKS5_PROXY).String()
		} else {
			cfg.ProxySettings.I2pSocks5Addr = "127.0.0.1:4447"
		}
	} else {
		panicNoSection(PROXY_SETTINGS)
	}

	if iniData.HasSection(NODE_SERVICE) {
		nodeServiceSection := iniData.Section(NODE_SERVICE)
		if nodeServiceSection.HasKey(ENABLED) {
			cfg.P2PServerEnabled, err = nodeServiceSection.Key(ENABLED).Bool()
			if err != nil {
				cfg.P2PServerEnabled = true
			}
		} else {
			cfg.P2PServerEnabled = true
		}
		if cfg.P2PServerEnabled {
			if nodeServiceSection.HasKey(HIDDEN_SERVICE) {
				hiddenService := nodeServiceSection.Key(HIDDEN_SERVICE).String()
				if hiddenService == "tor" {
					serviceProto := &hiddenservice.TorProto{}

					if nodeServiceSection.HasKey(TOR_CONTROL_PORT) {
						serviceProto.TorControlPort, err = nodeServiceSection.Key(TOR_CONTROL_PORT).Int()
						if err != nil {
							serviceProto.TorControlPort = 9051
						}
						if nodeServiceSection.HasKey(TOR_CONTROL_AUTH_PASSWORD) {
							serviceProto.TorControlPassword = nodeServiceSection.Key(TOR_CONTROL_AUTH_PASSWORD).String()
						}
					}
					cfg.P2PServerProto = serviceProto
				} else if hiddenService == "i2p" {

					if nodeServiceSection.HasKey(I2P_SAM_PORT) {
						serviceProto := &hiddenservice.I2PProto{}

						serviceProto.SamAPIPort, err = nodeServiceSection.Key(I2P_SAM_PORT).Int()
						if err != nil {
							serviceProto.SamAPIPort = 7656
						}
						if nodeServiceSection.HasKey(I2P_SAM_USER) {
							serviceProto.NeedAuth = true
							serviceProto.SamUser = nodeServiceSection.Key(I2P_SAM_USER).String()
							if nodeServiceSection.HasKey(I2P_SAM_PASSWORD) {
								serviceProto.SamPassword = nodeServiceSection.Key(I2P_SAM_PASSWORD).String()
							} else {
								panicNoKey(I2P_SAM_PASSWORD)
							}
						} else {
							serviceProto.NeedAuth = false
						}
						cfg.P2PServerProto = serviceProto
					}
				}
			} else {
				serviceProto := &hiddenservice.IP4TCPProto{}
				if nodeServiceSection.HasKey(BIND_ADDRESS) {
					serviceProto.BindAddress = nodeServiceSection.Key(BIND_ADDRESS).String()
				} else {
					serviceProto.BindAddress = "127.0.0.1:4222"
				}
				cfg.P2PServerProto = serviceProto
			}
		}
	} else {
		panicNoSection(NODE_SERVICE)
	}
	return cfg

}
